src/query_engine.js

```diff
@@ -576,12 +577,18 @@ QueryEngine.prototype.denormalizeBindings = function(bindings, env, callback) {
                 bindings[variable] = envOut[oid];
                 k();
             } else {
-                that.lexicon.retrieve(oid, function(val){
-                    bindings[variable] = val;
+                var cb = function(val){
+                    lazyCache['#'+oid] = bindings[variable] = val;
                     if(val.token === 'blank')
                         env.blanks[val.value] = oid;
                     k();
-                });
+                };
+                if (lazyCache['#'+oid]) {
+                    process.nextTick(function() { cb(lazyCache['#'+oid]) });
```

Comment:
Okay, I'll change it. And add the same improvement at a similar place (three times better performance in my tests).